### PR TITLE
[node:stream] add iterator, all the Array-like Readable methods

### PR DESCRIPTION
**Pull request by Touffy**
Declares [readable.iterator](https://nodejs.org/dist/latest-v18.x/docs/api/stream.html#readableiteratoroptions) and all the Array-like methods such as [map](https://nodejs.org/dist/latest-v18.x/docs/api/stream.html#readablemapfn-options) and [reduce](https://nodejs.org/dist/latest-v18.x/docs/api/stream.html#readablereducefn-initial-options) that were added to NodeJS Readable streams in between v16.3 and v17.5.

I've added generics where it could do some good (just `find` and `reduce`, as it happens). But as usual with NodeJS streams, it's a whole lot of `any` everywhere else.

### CODE REVIEW

**Comment by Touffy** [comment]
So… looks like a couple packages have incorrect typings that conflict with Node's built-in Readable methods :

### readable-stream

(third-party Readable implementation, obviously didn't get the memo about those new methods)

Hopefully, readable-stream will vanish into obscurity or get around to implementing all of modern Readables. It's not like we don't have native streams in the browser AND Node.js nowadays. In the meantime, the new methods are `Omit`ted when their test attempt to validate that the library's Readables are assignable to Node Readables.

### jsforce

(depends on readable-stream and overrides Readable map/filter)

I had no choice but to `@ts-ignore` the overrides in jsforce, since their implementation is incompatible with the built-in methods. Sucks for them, but then again, it's already a SalesForce utility.

**Review by mcollina — APPROVED** [review]
lgtm

**Comment by DangerBotOSS** [comment]
Inspecting the JavaScript source for this package found some properties that are not in the .d.ts files.
The check for missing properties isn't always right, so take this list as advice, not a requirement.
## jsforce ([<kbd>unpkg</kbd>](https://unpkg.com/browse/jsforce@latest/))
was missing the following properties:
1. addListener
1. on
1. once
1. removeListener
1. off

<details>
<summary>as well as these 12 other properties...</summary>
<p>removeAllListeners, setMaxListeners, getMaxListeners, listeners, rawListeners, emit, listenerCount, prependListener, prependOnceListener, eventNames, VERSION, require</p>
</details>

<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 260c132b75a4f765b5a4269fc76a9b2de3a9da4d
</p>

**Review by Semigradsky — COMMENTED** [review]
Great work!

**Review by dolanmiu — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]
@Touffy thanks!

**Comment by Touffy** [comment]
Sorry for the late change.

It turns out that `readable-stream` already implements all the methods, but — just like Node — they were not typed. So I have taken this opportunity to add the types to `readable-stream`, which makes its `Readable` once more assignable to Node `Readable` (there's still a small issue with `ReadableOptions` that should be addressed in another PR).

Thanks to that, there is no more need to skip a bunch of tests in `rdf-dataset-ext`. So we have better typing for `readable-stream`, and one fewer affected package !

I also made a small improvement to `Readable.reduce` to prioritise the inference of its return type when the reducer callback is loosely typed without an initial value, and added tests for it and `Readable.find`. Those changes are in a separate commit to help reviewers.

**Review by Semigradsky — APPROVED** [review]

File: types/jsforce/query.d.ts
Comment (andrewbranch) [inline types/jsforce/query.d.ts:26]: Perhaps this could be represented better, and without `ts-ignore`, by making an interface that `extends Omit<Readable, 'map' | 'filter'>` and then declaring a variable for the `Query` constructor, with a construct signature that returns that `Query` interface (the pattern that DOM types use in lib.dom.d.ts). But I think this does have the desired effect, so I’ll leave it up to others whether to revisit this.

**Review by andrewbranch — APPROVED** [review]

**Review by peterblazejewicz — APPROVED** [review]

File: types/jsforce/query.d.ts
Comment (Touffy) [reply types/jsforce/query.d.ts:26]: AFAIK, you can't declare a class that `extends` a type. It has to be another class (a value, not a type).

You *could* replace the `extends` with `implements` and use `Omit` on the Readable interface, but it doesn't mean the same thing. Particularly where `instanceof` is concerned, and the prototype chain. I haven't looked at the jsforce implementation so I don't know if that's more or less accurate, but if the source code has an actual `extends Readable`, then the type should have it too.

Using `implements` and `Omit` would avoid the `@ts-ignore`, but I think it's more useful to keep the `instanceof Readable` working. A lot of stream consumers want an actual Readable as input, not just "something that acts like a stream", and I would bet it's preferable to keep (slightly) lying to them than to accommodate TypeScript's narrow-minded interpretation of `extends`.

I'm going to merge as-is, and if the jsforce maintainers want to deal with this another way, they can handle the consequences themselves.

**Comment by Touffy** [comment]
Ready to merge

File: types/jsforce/query.d.ts
Comment (andrewbranch) [reply types/jsforce/query.d.ts:26]: jsforce is implemented with an ES5-style class-like function and `util.inherits`. But this isn’t particularly relevant for types in regard to `instanceof`. My suggestion is something like this:

```ts
declare const Query: QueryConstructor;
interface Query<T> extends Omit<Readable, 'map' | 'filter'>, Promise<T> {
  // ...
}
interface QueryConstructor {
  new<T>(/* whatever the constructor args are */): Query<T>
}
```

which I think preserves all the behavior you’d want.

Comment (Touffy) [reply types/jsforce/query.d.ts:26]: `utils.inherits` has the same runtime effect as `extends`, right ? (except a Symbol maybe ?) It's just that TypeScript chooses to interpret the former as "nothing to see here" and the latter as "Oh ! I saw this in Java, it must be the same thing !"

My point is, the `@ts-ignore` is just working around a TypeScript design issue, not an actual error in the types. The output (what really matters to users) is correct : you have something that has Readable.prototype in its prototype chain (which can only be expressed by `extends Readable`, I think), and its `filter` and `map` methods are typed correctly (you just have to force TypeScript to let you override them in a way that isn't classically extending the superclass).

Comment (andrewbranch) [reply types/jsforce/query.d.ts:26]: > (which can only be expressed by extends Readable, I think)

```diff
interface QueryConstructor {
  new<T>(/* whatever the constructor args are */): Query<T>
+ prototype: Query;
}
```

This is what all the class-like objects in lib.d.ts do. For example:

```ts
interface StringConstructor {
    new (value?: any): String;
    (value?: any): string;
    prototype: String;
    fromCharCode(...codes: number[]): string;
}

/**
  * Allows manipulation and formatting of text strings and determination and location of substrings within strings.
  */
declare var String: StringConstructor;
```

Comment (Touffy) [reply types/jsforce/query.d.ts:26]: Nice. I had never paid attention to the `prototype` there.

Sadly, TypeScript doesn't care much about that `prototype`. For example :
```ts
function f(thing: Query | Readable) {
  if (thing.constructor === Readable) {
    // `thing` is not narrowed to a Readable here :(
  }
  if (thing instanceof Readable) return
  // `thing` is not narrowed to `never` after that, even though this is in fact unreachable.
  // The only way to get `never` here is still `class Query extends Readable` as I suspected.
}
```

Using the @ts-ignores wouldn't necessarily lead to errors if some code uses `instanceof Readable` before calling `find` or `map` because in that case, `instanceof Readable` does *not* narrow `Readable | Query` to just `Readable`. So I still prefer my version, even though ideally, authors should rely less on `instanceof` when a narrower check is possible (like checking for `"then" in thing` instead of `thing instanceof Promise`). Symbols really help with that, but it's too late for Streams.

**Pull request merged**
